### PR TITLE
Use RabbitmqCluster namespace in RabbitmqClusterReference

### LINKED ISSUE #454: Multiple Knative Brokers in different namespaces that connect and use a single RabbitMQ cluster instance
**Problem**
A user might prefer to create a single RabbitMQ cluster with sufficient capacity to service all messaging needs of a single K8s cluster. It would be desirable to have multiple Knative Brokers (across multiple namespaces such as foo, bar) connect to and use a single RabbitMQ cluster thats created in baz namespace.

**[Persona:](https://github.com/knative/eventing/blob/main/docs/personas.md)**
Which persona is this feature for?
This feature would be used by the K8s Cluster Administrator

**Exit Criteria**
Be able to setup something like
1. Install RabbitMQ operators and eventing-rabbitmq
2. Create RabbitMQ Cluster `my-rabbit-cluster` in namespace `baz`
2. Create a Knative Broker `foo-broker` (backed by `my-rabbit-cluster`) in namespace `foo`
3. Create a Knative Broker `bar-broker` (backed by `my-rabbit-cluster`) in namespace `bar`
4. Ensure the status of both Brokers is `true`
5. Ensure that an event producer (eg. PingSource) can send events to the brokers and an event consumer (eg. event-display) can consume events from the two brokers.

### PR DESCRIPTION
**Pull request by gvmw**
# Changes

- :bug: &nbsp; Use the RabbitmqCluster namespace in cluster references, otherwise resources will fail to reconcile when RabbitMQ is running in a different namespace

When declaring a new `Broker` of class `RabbitMQBroker`, use the `RabbitmqCluster` namespace in `RabbitmqClusterReference`, otherwise the dependent objects which get declared - i.e. `Exchange`, `Queue`, `Binding` - and are expected to be reconciled by the https://github.com/rabbitmq/messaging-topology-operator will fail to be created.

This has more details: https://github.com/knative-sandbox/eventing-rabbitmq/issues/454

/kind bug

Fixes #454

### CODE REVIEW

**Comment by gvmw** [comment]
We must have a missing test, because the new code fails to create dependent `Exchange` resources, as outlined above.

The other thing which I've noticed is that I have a lot of changes in my working copy, and I suspect that I shouldn't have run `go mod vendor`. Does this look right?

```sh
On branch add-support-for-single-rabbitmq-instance
Your branch is up to date with 'origin/add-support-for-single-rabbitmq-instance'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   go.mod
        modified:   go.sum
        modified:   vendor/modules.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        vendor/github.com/google/mako/go/quickstore/BUILD
        vendor/github.com/google/mako/internal/go/common/BUILD
        vendor/github.com/grpc-ecosystem/grpc-gateway/internal/BUILD.bazel
        vendor/github.com/grpc-ecosystem/grpc-gateway/runtime/BUILD.bazel
        vendor/github.com/grpc-ecosystem/grpc-gateway/utilities/BUILD.bazel
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/binding_webhook.go
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/exchange_webhook.go
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/federation_webhook.go
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/permission_webhook.go
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/policy_webhook.go
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/queue_webhook.go
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/schemareplication_webhook.go
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/shovel_webhook.go
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/user_webhook.go
        vendor/github.com/rabbitmq/messaging-topology-operator/api/v1beta1/vhost_webhook.go
        vendor/k8s.io/apimachinery/pkg/api/errors/OWNERS
        vendor/k8s.io/apimachinery/pkg/api/meta/OWNERS
        vendor/k8s.io/apimachinery/pkg/api/resource/OWNERS
        vendor/k8s.io/apimachinery/pkg/apis/meta/v1/OWNERS
        vendor/k8s.io/apimachinery/pkg/util/mergepatch/OWNERS
        vendor/k8s.io/apimachinery/pkg/util/strategicpatch/OWNERS
        vendor/k8s.io/apimachinery/third_party/forked/golang/json/OWNERS
        vendor/k8s.io/client-go/pkg/apis/clientauthentication/OWNERS
        vendor/k8s.io/client-go/plugin/pkg/client/auth/gcp/OWNERS
        vendor/k8s.io/client-go/rest/OWNERS
        vendor/k8s.io/client-go/tools/auth/OWNERS
        vendor/k8s.io/client-go/tools/cache/OWNERS
        vendor/k8s.io/client-go/tools/leaderelection/OWNERS
        vendor/k8s.io/client-go/tools/metrics/OWNERS
        vendor/k8s.io/client-go/tools/record/OWNERS
        vendor/k8s.io/client-go/transport/OWNERS
        vendor/k8s.io/client-go/util/cert/OWNERS
        vendor/k8s.io/client-go/util/keyutil/OWNERS
        vendor/k8s.io/client-go/util/retry/OWNERS
        vendor/k8s.io/code-generator/OWNERS
        vendor/k8s.io/code-generator/cmd/client-gen/OWNERS
        vendor/k8s.io/code-generator/cmd/go-to-protobuf/OWNERS
        vendor/k8s.io/component-base/config/OWNERS
        vendor/k8s.io/klog/OWNERS
        vendor/k8s.io/klog/v2/OWNERS
        vendor/k8s.io/kube-openapi/pkg/generators/rules/OWNERS
        vendor/k8s.io/kube-openapi/pkg/util/proto/OWNERS
        vendor/k8s.io/utils/pointer/OWNERS
        vendor/knative.dev/eventing/test/performance/infra/OWNERS
        vendor/knative.dev/eventing/test/test_images/performance/OWNERS
        vendor/knative.dev/hack/OWNERS
        vendor/knative.dev/hack/OWNERS_ALIASES
        vendor/knative.dev/pkg/apis/OWNERS
        vendor/knative.dev/pkg/apis/duck/OWNERS
        vendor/knative.dev/pkg/configmap/OWNERS
        vendor/knative.dev/pkg/controller/OWNERS
        vendor/knative.dev/pkg/hack/OWNERS
        vendor/knative.dev/pkg/hash/OWNERS
        vendor/knative.dev/pkg/injection/OWNERS
        vendor/knative.dev/pkg/kflag/OWNERS
        vendor/knative.dev/pkg/kmeta/OWNERS
        vendor/knative.dev/pkg/logging/OWNERS
        vendor/knative.dev/pkg/metrics/OWNERS
        vendor/knative.dev/pkg/network/OWNERS
        vendor/knative.dev/pkg/reconciler/OWNERS
        vendor/knative.dev/pkg/resolver/OWNERS
        vendor/knative.dev/pkg/test/OWNERS
        vendor/knative.dev/pkg/test/mako/OWNERS
        vendor/knative.dev/pkg/tracing/config/OWNERS
        vendor/knative.dev/pkg/webhook/OWNERS
        vendor/sigs.k8s.io/controller-runtime/OWNERS
        vendor/sigs.k8s.io/controller-runtime/OWNERS_ALIASES
        vendor/sigs.k8s.io/yaml/OWNERS
```

**Comment by gvmw** [comment]
I've realised the issues with my initial approach. If resources belonging to a single namespace start getting spread, the logical namespace encapsulation gets broken and we would end up with a mess. Resources that should be close to one another aren't, which makes lifecycle events more complicated - especially deletions.

It would make a lot more sense to define the `namespace` property in `rabbitmqClusterReference` instead. This seems to be supported:

https://github.com/rabbitmq/messaging-topology-operator/blob/51f6d88b214e17b12dd31cd6b4c555c1531b3f0c/internal/cluster_reference.go#L26

I have confirmed it in the type definition too:

https://github.com/rabbitmq/messaging-topology-operator/blob/51f6d88b214e17b12dd31cd6b4c555c1531b3f0c/api/v1beta1/queue_types.go#L46-L54

Let's see what happens with this approach.

**Comment by gvmw** [comment]
I think that I'm going in the wrong direction with deps. Current status:

```go
go test -race knative.dev/eventing-rabbitmq/pkg/reconciler/broker
# knative.dev/pkg/client/injection/kube/client
vendor/knative.dev/pkg/client/injection/kube/client/client.go:1236:9: cannot use &wrapAppsV1DeploymentImpl{...} (type *wrapAppsV1DeploymentImpl) as type "k8s.io/client-go/kubernetes/typed/apps/v1".DeploymentInterface in return argument:
        *wrapAppsV1DeploymentImpl does not implement "k8s.io/client-go/kubernetes/typed/apps/v1".DeploymentInterface (missing ApplyScale method)
vendor/knative.dev/pkg/client/injection/kube/client/client.go:1253:5: cannot use (*wrapAppsV1DeploymentImpl)(nil) (type *wrapAppsV1DeploymentImpl) as type "k8s.io/client-go/kubernetes/typed/apps/v1".DeploymentInterface in assignment:
        *wrapAppsV1DeploymentImpl does not implement "k8s.io/client-go/kubernetes/typed/apps/v1".DeploymentInterface (missing ApplyScale method)
vendor/knative.dev/pkg/client/injection/kube/client/client.go:1400:5: cannot use (*wrapAppsV1ReplicaSetImpl)(nil) (type *wrapAppsV1ReplicaSetImpl) as type "k8s.io/client-go/kubernetes/typed/apps/v1".ReplicaSetInterface in assignment:
        *wrapAppsV1ReplicaSetImpl does not implement "k8s.io/client-go/kubernetes/typed/apps/v1".ReplicaSetInterface (missing ApplyScale method)
vendor/knative.dev/pkg/client/injection/kube/client/client.go:1547:5: cannot use (*wrapAppsV1StatefulSetImpl)(nil) (type *wrapAppsV1StatefulSetImpl) as type "k8s.io/client-go/kubernetes/typed/apps/v1".StatefulSetInterface in assignment:
        *wrapAppsV1StatefulSetImpl does not implement "k8s.io/client-go/kubernetes/typed/apps/v1".StatefulSetInterface (missing ApplyScale method)
vendor/knative.dev/pkg/client/injection/kube/client/client.go:2697:5: cannot use (*wrapAppsV1beta2StatefulSetImpl)(nil) (type *wrapAppsV1beta2StatefulSetImpl) as type "k8s.io/client-go/kubernetes/typed/apps/v1beta2".StatefulSetInterface in assignment:
        *wrapAppsV1beta2StatefulSetImpl does not implement "k8s.io/client-go/kubernetes/typed/apps/v1beta2".StatefulSetInterface (missing ApplyScale method)
vendor/knative.dev/pkg/client/injection/kube/client/client.go:6049:5: cannot use (*wrapCoreV1PodImpl)(nil) (type *wrapCoreV1PodImpl) as type "k8s.io/client-go/kubernetes/typed/core/v1".PodInterface in assignment:
        *wrapCoreV1PodImpl does not implement "k8s.io/client-go/kubernetes/typed/core/v1".PodInterface (missing EvictV1 method)
vendor/knative.dev/pkg/client/injection/kube/client/client.go:6170:112: undefined: "k8s.io/api/core/v1".EphemeralContainers
vendor/knative.dev/pkg/client/injection/kube/client/client.go:6174:90: undefined: "k8s.io/api/core/v1".EphemeralContainers
vendor/knative.dev/pkg/client/injection/kube/client/client.go:7808:5: cannot use (*wrapExtensionsV1beta1DeploymentImpl)(nil) (type *wrapExtensionsV1beta1DeploymentImpl) as type "k8s.io/client-go/kubernetes/typed/extensions/v1beta1".DeploymentInterface in assignment:
        *wrapExtensionsV1beta1DeploymentImpl does not implement "k8s.io/client-go/kubernetes/typed/extensions/v1beta1".DeploymentInterface (missing ApplyScale method)
vendor/knative.dev/pkg/client/injection/kube/client/client.go:8368:5: cannot use (*wrapExtensionsV1beta1ReplicaSetImpl)(nil) (type *wrapExtensionsV1beta1ReplicaSetImpl) as type "k8s.io/client-go/kubernetes/typed/extensions/v1beta1".ReplicaSetInterface in assignment:
        *wrapExtensionsV1beta1ReplicaSetImpl does not implement "k8s.io/client-go/kubernetes/typed/extensions/v1beta1".ReplicaSetInterface (missing ApplyScale method)
vendor/knative.dev/pkg/client/injection/kube/client/client.go:1236:9: too many errors
FAIL    knative.dev/eventing-rabbitmq/pkg/reconciler/broker [build failed]
```

After https://github.com/knative-sandbox/eventing-rabbitmq/pull/492/commits/9db13691d53d6aa0e991ff2d8e1b829006adfe8b, I think the update to `k8s.io/code-generator`, specifically https://github.com/knative-sandbox/eventing-rabbitmq/pull/492/commits/9db13691d53d6aa0e991ff2d8e1b829006adfe8b#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L31-R32 broke the `k8s.io/client-go`. I'm checking if any of the two topology operators depend on a newer client-go, and how to solve this.

**Comment by gvmw** [comment]
This is the commit that bumped `k8s.io/code-generator` to `0.22.0`: https://github.com/rabbitmq/messaging-topology-operator/commit/2765a03aeaf3820744688e56a7ee4ab43b7fd392 . It may be unrelated, but it's the first dependency that introduces `0.22` in this mix, and I'm assuming that makes everything bump.

I was able to lock it to a specific version by running `go get k8s.io/code-generator@v0.21.4`, but that has also downgraded the messaging-topology-operator. I think that I'm missing a trick here, will ask for help.

**Comment by gvmw** [comment]
I've made progress with this, there is a change that needs to happen in messaging-topology-operator:  https://github.com/rabbitmq/messaging-topology-operator/discussions/256

**Comment by gvmw** [comment]
How do I fix the tide error @benmoss?

The conflict is showing me an old diff, I have since force pushed a few times and everything applies cleanly. If you run a git diff local, it will look different to the tide output above 🤷🏼‍♂️

**Comment by gvmw** [comment]
I am leaving a note for my future self. In case I come across this failure in the future:

```go
panic: /ko-app/broker
  flag redefined: kubeconfig

goroutine 1 [running]:
flag.(*FlagSet).Var(0xc000090720,
  {0x1eaddf0, 0xc0001b4570}, {0x1c48cde, 0xa}, {0x1c9322f, 0x36})
    flag/flag.go:879
  +0x2f4
flag.(*FlagSet).StringVar(...)
    flag/flag.go:762
knative.dev/pkg/environment.(*ClientConfig).InitFlags(0xc0001b4540,
  0xc000090720)
    knative.dev/pkg@v0.0.0-20211109100843-91d1932616a7/environment/client_config.go:45
  +0x134
knative.dev/pkg/injection.ParseAndGetRESTConfigOrDie()
    knative.dev/pkg@v0.0.0-20211109100843-91d1932616a7/injection/config.go:32
  +0x36
knative.dev/pkg/injection/sharedmain.MainWithContext({0x1eca468,
  0xc000362030}, {0x1c5ef4e, 0x1a}, {0xc0006bff60, 0x2, 0x2})
    knative.dev/pkg@v0.0.0-20211109100843-91d1932616a7/injection/sharedmain/main.go:125
  +0x78
knative.dev/pkg/injection/sharedmain.Main({0x1c5ef4e, 0x1a}, {0xc0006bff60,
  0x2, 0x2})
    knative.dev/pkg@v0.0.0-20211109100843-91d1932616a7/injection/sharedmain/main.go:105
  +0x8f
main.main()
    knative.dev/eventing-rabbitmq/cmd/controller/broker/main.go:29
  +0x4e
```

Remember that it is related to dependencies being updated, and `./hack/update-codegen.sh` not running.
The fix is to run it, then commit and push all the changes.

**Comment by gvmw** [comment]
The only thing missing at this point is an e2e test. Opening this up for review while I work on that.

File: pkg/reconciler/broker/broker.go
Comment (benmoss) [inline pkg/reconciler/broker/broker.go:301]:
```suggestion
			ClusterName:    b.Spec.Config.Name,
			ClusterNamespace:    b.Spec.Config.Namepace,
```

I was hoping to keep these APIs at a different level of abstraction by not including objects like the Broker and the Trigger. Like they should only know "RabbitMQ"-y things.

File: pkg/reconciler/trigger/resources/binding.go
Comment (benmoss) [inline pkg/reconciler/trigger/resources/binding.go:91]: Why these changes?

File: pkg/reconciler/trigger/resources/queue.go
Comment (benmoss) [inline pkg/reconciler/trigger/resources/queue.go:72]: Same, why make these changes?

**Review by benmoss — CHANGES_REQUESTED** [review]

File: pkg/reconciler/broker/broker.go
Comment (gvmw) [reply pkg/reconciler/broker/broker.go:301]: OK, I will pass in `ClusterName` & `ClusterNamespace` instead of `Broker`.

File: pkg/reconciler/trigger/resources/binding.go
Comment (gvmw) [reply pkg/reconciler/trigger/resources/binding.go:91]: `ExchangeLabels`, `QueueLabels` and `BindingLabels` are identical functions.

`ExchangeLabels` had one implementation, while `QueueLabels` and `BindingLabels` had another. This was my first step to unify them, as a precursor to maybe extracting into a single place.

I can revert if you think doing nothing is better than this change.

File: pkg/reconciler/trigger/resources/queue.go
Comment (gvmw) [reply pkg/reconciler/trigger/resources/queue.go:72]: https://github.com/knative-sandbox/eventing-rabbitmq/pull/492#discussion_r750154653

**Comment by gvmw** [comment]
@benmoss how can we change the label from `kind/enhancement` to `kind/bug`?

Even though I have changed the description since opening the PR, the label has not updated, which I think is the responsibility of https://github.com/knative-prow-robot

**Comment by gvmw** [comment]
I am starting to shave another yak: `make test-e2e`. Current status:

image

I don't know if you want to wait for me to do that, but until I can run e2e tests easily and repeatedly, it's impossible to iterate on the e2e test via GitHub Actions. I've checked the existing e2e tests, it will take me a few hours of experimenting to pick them up. This requires me to be able to run the easily and repeatedly, which brings me back to `make test-e2e`.

File: pkg/reconciler/trigger/resources/binding.go
Comment (benmoss) [reply pkg/reconciler/trigger/resources/binding.go:91]: Let's do this in a separate PR

**Comment by gvmw** [comment]
So that added the `kind/bug` label, nice!

How do we remove the `kind/enhancement` label?

File: pkg/reconciler/trigger/resources/binding.go
Comment (benmoss) [inline pkg/reconciler/trigger/resources/binding.go:40]: This field is unused

File: pkg/reconciler/trigger/resources/queue.go
Comment (benmoss) [inline pkg/reconciler/trigger/resources/queue.go:33]: This field is unused

**Review by benmoss — CHANGES_REQUESTED** [review]
Let's not change the label logic here (I opened #513), there's already a lot going on here

**Comment by benmoss** [comment]
https://prow.k8s.io/command-help?repo=knative-sandbox%2Feventing-rabbitmq
/remove-kind enhancement

**Comment by benmoss** [comment]
I'm using [this](https://gist.github.com/benmoss/5ad02fafc49880863f82c9e5f07f56e7) topology and seeing this error on the exchange:
```
Resource is not allowed to reference defined cluster reference. Check the namespace of the resource is allowed as part of the cluster's `rabbitmq.com/topology-allowed-namespaces` annotation
```

**Comment by benmoss** [comment]
Ah, I see https://github.com/knative-sandbox/eventing-rabbitmq/issues/454#issuecomment-943253640 references this annotation, its required for cross-namespace references

**Comment by benmoss** [comment]
So this is still WIP, waiting on the e2e test, right?

**Comment by gvmw** [comment]
> So this is still WIP, waiting on the e2e test, right?

Yes, that's right. Making slow progress, but getting there. I will commit what I have today (most of it is the `Makefile` which belongs in a different PR I think). Based on how today goes, it may help to pair-up as right now progress feels too slow and it's all down to learning and figuring out how the entire e2e fits together.

**Comment by gvmw** [comment]
> So this is still WIP, waiting on the e2e test, right?

I just remembered that we may want to wait for a messaging-topology release that has the feature which we depend on: https://github.com/rabbitmq/messaging-topology-operator/pull/257

WDYT @benmoss ?

**Comment by benmoss** [comment]
For https://github.com/rabbitmq/messaging-topology-operator/pull/257 I think it's fine to just pin our dependency to the commit, the released version is fine for users to install

**Comment by gvmw** [comment]
This is now ready for review. Everything that we need is now in, including end-to-end tests.

**Comment by gvmw** [comment]
After this is merged, remember to update https://github.com/knative/docs/blob/ad6446431ea4de30bcbf94e8a36bdf43df11943d/docs/eventing/broker/rabbitmq-broker/README.md?plain=1#L92

With the following:

```yaml
      annotations:
        # This enables a single RabbitMQ cluster to service multiple Brokers
        rabbitmq.com/topology-allowed-namespaces: "*"
```

**Comment by gvmw** [comment]
@embano1 I just checked and nightly builds have picked this change. Let us know if `gcr.io/knative-nightly/knative.dev/eventing-rabbitmq/cmd/controller/broker@sha256:91907b32bb4f1a1e883b87efdeca9f5c7513dfe38b958086a47283ab65566401` image works for you:

```sh
kubectl apply -f https://storage.googleapis.com/knative-nightly/eventing-rabbitmq/latest/rabbitmq-broker.yaml
```

**Comment by gvmw** [comment]
Had confirmation this morning from @embano1 that https://github.com/knative-sandbox/eventing-rabbitmq/pull/492#issuecomment-979073100 must go in as soon as https://github.com/knative/docs/pull/4518 gets merged.

**Comment by gvmw** [comment]
This should help: https://github.com/knative/docs/pull/4532

**Pull request merged**
